src/libcore/num/mod.rs

```diff
@@ -2250,7 +2250,8 @@ assert_eq!(
 );
 ```"),
             #[stable(feature = "int_to_from_bytes", since = "1.32.0")]
-            #[rustc_const_unstable(feature = "const_int_conversion", issue = "53718")]
+            #[rustc_const_stable(feature = "const_int_conversion", since = "1.43.0")]
+            #[allow_internal_unstable(const_transmute)]
             #[inline]
             pub const fn to_ne_bytes(self) -> [u8; mem::size_of::<Self>()] {
                 // SAFETY: integers are plain old datatypes so we can always transmute them to
```

Comment:
I would use the union based implementation as done in the aforementioned PR  instead of the transmute based one with the macro, as that seems more hacky.

Reply:
@Centril This code doesn't compile. Transmute or not you'll either need that hack macro or fix `#[allow_internal_unstable]`.

```rust
#[allow_internal_unstable(const_fn)] // <-- no effect
const fn h(b: [u8; 8]) -> u64 {
    union U {
        a: u64,
        b: [u8; 8],
    }
    let u = U { b };
    unsafe {
        u.a //~ ERROR [E0723]: accessing union fields is unstable
    }
}
```

Reply:
The macro doesn't do the same thing as `#[allow_internal_unstable]` on a `const fn`, at all.

The `const fn` is a different feature that's reusing the attribute and given how confused I was by it, it should be using a different attribute.

Reply:
Whether "fixing `#[allow_internal_unstable]`" or "create another attribute for `const fn`" requires changing the compiler and blocks this PR.

Reply:
Well, changing the attribute's name to not be confusing won't make it work for non-language features anyway.

My point was that there's no reason to use a macro, because it's the same as adding the feature at the crate level, just unnecessarily complicated.
It doesn't have the special semantics @oli-obk added to the attribute when applied to `const fn`, it's just a free pass.

Overall I am pretty confused at how this system is *supposed to* work and I don't think I want to merge this PR until we hear from @oli-obk.

Also, we don't need to wait for a change to the attribute to trickle into beta because we can always `#![cfg_attr(bootstrap, feature(const_transmute))]` or w/e.

Reply:
Given the conceptual similarities between macros and `const fn` (in the widest sense, "stuff that is executed during compile time"), I actually liked that it re-used the same attribute. It's doing conceptually similar things: locally control the use of unstable features.

If `allow_internal_unstable` would have been introduced to apply to macros and `const fn` from the start, what would be confusing about it? Or is the only confusion that an attribute that you thought did a particular thing, was expanded to also do a very similar thing elsewhere?

Reply:
But they do very different things?

The macro feature is a hack around the lack of stability hygiene (although I guess nowadays it's stricter than real stability hygiene would be).
We just use it whenever we have unstable implementation details of a macro.

Whereas the `const fn` use of the attribute only applies to *stably* `const fn` functions, and it's for controlling which `const_*` language features we expose as stable. We don't slap it on "as needed" like the macro usecase (quite the opposite!), and it barely does anything overall.

This is why I think the latter deserves its own attribute with its own checking that disallows using it when it doesn't do anything, to alleviate some of the confusion here.
It would also allow searching for it more easily, for independently doing a "quick audit" of the features it's used with.

Reply:
>  it's for controlling which const_* language features we expose as stable

At least to someone foreign with the details like me, that sounds a lot like "controlling which unstable features a macro exposes to stable code".

Reply:
> At least to someone foreign with the details like me, that sounds a lot like "controlling which unstable features a macro exposes to stable code".

1. the macro could very well be unstable, that changes nothing (unlike the `const fn` situation  which is only relevant for "stably `const fn`")
2. it's *not* for control, it's out of necessity, we don't have the necessary hygiene information to be able to do this without the attribute
3. the feature list in the attribute is a recent development, before that we wouldn't have thought twice about throwing the attribute out if we had gotten stability hygiene

If anything this probably shows we might want `#![feature(...)]` on arbitrary scopes, not just crate-level, and then you can lint against `const_*` features being applied at a scope larger than an individual `const fn`, if you want to.

`const fn` definitions then wouldn't need a separate attribute, since the tight feature-gate scoping would just be a general feature at that point that they can just use instead of something special for them.

While `macro_rules!` macros would continue to require an attribute, although it might make sense to have `#[feature(foo)] #[rustc_inherit_unstable]` instead of `#[allow_internal_unstable(foo)]`.

Anyway this is all just speculation, *if nothing else*, I want to use a different name that's clearly about `const`(`fn`) to make it searchable and more of a warning sign.

Reply:
I'm fine just naming it `#[rustc_allow_const_feature(foo)]`, but the reason I reused the existing attribute was because it seems like the same thing to me. There's some thing callable by users (stably or unstably) that uses an unstable feature inside and we don't want the user to have to specify the feature. It's slightly better isolated in the const fn case than in the macro case, but it's still essentially the same logic to me.

Anyway, as said at the start of this post, I'm fine renaming it. Especially since people are confused by const fn having the same attribute as macro_rules.

Reply:
I converted from transmute to union as suggested by @Centril, and that incidentally allowed this to build as unions are language features not library features.

However, now I'm reading the point by @kennytm that a union doesn't check the type size, so if that causes an issue I can revert that bit. (I don't think it's an issue here, the union is between `$SelfT` and `[u8; mem::size_of::<$SelfT>]`.)

Reply:
Yea the change looks alright. We'll want to move back to transmute at some point, so I added the const-hack label to this PR.

Reply:
> I converted from transmute to union as suggested by @Centril, and that incidentally allowed this to build as unions are language features not library features.

Fwiw, this was not incidental at all; that's why I suggested it. ;)